### ISSUE TITLE
need to getting list of externals in the same order that is defined in the config file

### ISSUE THREAD

**Issue by uturuncoglu** [ESMCI/manage_externals#147]
I am trying to checkout following config file that is created for the NOAA UFS application,

```
[externals_description]
schema_version = 1.0.0

[hafs]
branch = feature/hafs_nogsi
protocol = git
repo_url = https://github.com/hafs-community/HAFS.git
local_path = HAFS
required = True

[hycom]
# feature/NUOPC_merged
tag = 28b84ce
protocol = git
repo_url = https://github.com/hafs-community/HYCOM-src.git
local_path = HAFS/sorc/hafs_forecast.fd/HYCOM
required = True

[cime]
# master 
tag = 35fb2d7
protocol = git
repo_url = https://github.com/ESMCI/cime.git
local_path = cime
externals = ../Externals_cime.cfg
required = True

[cdeps]
# master 
tag = e99ec4f
protocol = git
repo_url = https://github.com/ESCOMP/CDEPS.git
local_path = components/cdeps
required = True

[fv3gfs_interface]
# ufs_update_11
tag = 6a78103
protocol = git
repo_url = https://github.com/ESCOMP/fv3gfs_interface.git
local_path = HAFS/sorc/hafs_forecast.fd/FV3/cime/
required = True

[nems_interface]
tag = 3c36830
protocol = git
repo_url = https://github.com/ESCOMP/NEMS_interface.git
local_path = HAFS/sorc/hafs_forecast.fd/NEMS/cime/
required = True
```

but I am getting error like following,

```
Processing externals description file : Externals.cfg
Checking status of externals: hycom, cime, fv3gfs_interface, hafs, cdeps, nems_interface, 
Checking out externals: hycom, cime, fv3gfs_interface, hafs, ERROR:root:Command '[u'git', u'checkout', u'--quiet', u'hafs-community_HAFS.git/feature/hafs_nogsi']' returned non-zero exit status 1
ERROR:root:Failed with output:
    error: Your local changes to the following files would be overwritten by checkout:
    	Externals.cfg
    Please commit your changes or stash them before you switch branches.
    Aborting

ERROR: In directory
    /glade/u/home/turuncu/UFS/ufs-hafs-app/HAFS
Process did not run successfully; returned status 1:
    git checkout --quiet hafs-community_HAFS.git/feature/hafs_nogsi
See above for output from failed command.

ERROR:root:Failed with output:
    error: Your local changes to the following files would be overwritten by checkout:
    	Externals.cfg
    Please commit your changes or stash them before you switch branches.
    Aborting
```

I think that the reason of getting this error is to checkout  `fv3gfs_interface` before checking out `hafs` because  `fv3gfs_interface` is checking out as a subfolder in HAFs model. Currently, there is no way to control the order of checking out externals in the order that are defined in the Externals.cfg file and that could be a nice feature. I try to look at the possible solution and it seems that the configuration file is read by **ConfigParser** module and I could not find any information to read with that way. I tried to sort the models but that also not working because the alphabetical order also causes same problem. So, I need explicit control of checking out externals. Any suggestion will be helpful at this point.

**Comment by billsacks** [comment]
I don't understand all of the needs and constraints of the UFS project, but I'll just chime in to say that I find it confusing in general when we have one external nested inside another external like this. I feel that things like this should be done by either:

(1) If you want all of the externals managed from the top level, then they should appear in separate directories at the top level, not having externals nested inside other externals.

OR

(2) If this directory organization is needed, it feels to me like it should be handled by having HAFS have its own Externals_HAFS file, leveraging the recursive capabilities of manage_externals.

Again, though, I realize that these suggestions may not be feasible given the constraints of the UFS project.

**Comment by uturuncoglu** [comment]
@billsacks Thanks for your help. I think that the second option could work for me. let me test that one.

**Comment by uturuncoglu** [comment]
@billsacks I created seperate externals for HAFS related modules as follows,

```
[hycom]
# feature/NUOPC_merged
tag = 28b84ce
protocol = git
repo_url = https://github.com/hafs-community/HYCOM-src.git
local_path = sorc/hafs_forecast.fd/HYCOM
required = True

[fv3gfs_interface]
# ufs_update_11
tag = 6a78103
protocol = git
repo_url = https://github.com/ESCOMP/fv3gfs_interface.git
local_path = sorc/hafs_forecast.fd/FV3/cime/
required = True

[nems_interface]
# master
tag = 3c36830
protocol = git
repo_url = https://github.com/ESCOMP/NEMS_interface.git
local_path = sorc/hafs_forecast.fd/NEMS/cime/
required = True

[externals_description]
schema_version = 1.0.0
```

and the hafs is defined like following,

```
[hafs]
branch = feature/hafs_nogsi
protocol = git
repo_url = https://github.com/hafs-community/HAFS.git
local_path = HAFS
externals = ../Externals_hafs.cfg
required = True
```

but in this case, when I try to checkout hafs I am loosing all the directories in the  HAFS/sorc/hafs_forecast.fd/ folder and I could see only the modules that are checked out by the Externals_hafs.cfg. Do you have any idea why?

**Comment by billsacks** [comment]
I think what you're saying is that you created an Externals_hafs.cfg file at the top level where you have defined these externals. Sorry, I wasn't very clear, but that's not what I meant: I find that pattern confusing as well. What I meant is that, if you're going to checkout externals inside the HAFS directory, then that implies to me that those are owned / managed by HAFS, and thus should be defined by an Externals_hafs.cfg file that lives *inside* the HAFS repository (at the top level of that repository). For example, see how CAM is defined in https://github.com/ESCOMP/CESM/blob/master/Externals.cfg, and how this ends up picking up the Externals_CAM.cfg file that is defined in CAM itself (https://github.com/ESCOMP/CAM/blob/cam_development/Externals_CAM.cfg).

I know that the pattern of using `../Externals_hafs.cfg` follows what is currently done for cime (using `../Externals_cime.cfg`), but @mvertens and I feel that this is problematic and we shouldn't be doing things that way.

That said, I'm not sure if this actually explains the problem you're seeing. I don't have time right now to dig into that.

**Comment by uturuncoglu** [comment]
@billsacks Unfortunately, we have no any control in HAFS repository and it is hard to put Externals_hafs.cfg inside of HAFS app and it will also create a nightmare for us in terms of managing that file. I am not sure what we need to do at this point.

**Comment by uturuncoglu** [comment]
@billsacks enforcing the order of checkout and putting all HAFS related externals to main Externals.cfg could solve the problem. I could test or use it only in my app if there is a way to do it.

**Comment by gold2718** [comment]
@billsacks, what is the problem with having the Externals_HAFS.cfg at the top level?
This idiom is being actively used, for instance in [MOM_interface](https://github.com/ESCOMP/MOM_interface).

**Comment by billsacks** [comment]
I guess I'm coming to see that there are situations where this idiom may be required, namely: where the sub-externals really need to appear nested inside some other external `foo`, and where we don't have control over `foo` so can't add or maintain an Externals file inside `foo`. It sounds like that really is the situation here, and maybe it's the situation for `MOM_interface` as well. So in this case, maybe the right thing to do is to use this idiom, which would mean resolving the issue @uturuncoglu raised here https://github.com/ESMCI/manage_externals/issues/147#issuecomment-685892308 (which, sorry, I don't have time to help with this week).

My comment perhaps came as too much of a knee-jerk reaction that I developed from seeing the use of this idiom for managing CMEPS and CDEPS in Externals_CIME files: In this case, I found it confusing to get my head around what was going on at first ("okay, so top level CESM gets cime, then bounces back up to the top level in order to get some other externals from a different file which it then clones inside the cime repository...."). In this case, I felt like CMEPS and CDEPS could be managed directly from the main Externals file rather than an Externals_cime, and could be made to appear outside of cime – so that we avoid having CESM place externals inside other externals, which just felt confusing and weird to me.

**tl;dr: I feel like we should think twice before using this idiom (having, e.g., Externals_HAFS.cfg at the top level), but if it's truly needed – which it does seem to be in this case – then I drop my objection.**

**Comment by jedwards4b** [comment]
Why can we not impose an order on the Externals.cfg file - this would eliminate a lot of questionable hacks that we are having to do.

**Comment by uturuncoglu** [comment]
@jedwards4b +1 from me

**Comment by gold2718** [comment]
> Why can we not impose an order on the Externals.cfg file - this would eliminate a lot of questionable hacks that we are having to do.

I believe you will have to change some of manage_externals' internal data structures to maintain an ordering (I think in the current implementation, stuff comes out in hash order).

**Comment by jedwards4b** [comment]
I would take this on if I understood the code flow a little better.  Where is the external file read and where is the loop over repos?

**Comment by gold2718** [comment]
> ... this would eliminate a lot of questionable hacks that we are having to do.

Can you specify these 'hacks'? It seems to me that this is a proper use of hierarchical externals files to match a hierarchical source tree. The externals file has to be at a higher level due to ownership issues, the alternative is to make a fork of the repo just to add an externals file.

**Comment by jedwards4b** [comment]
An example hack is found in https://github.com/ESCOMP/UFSCOMP.git hash efb5d6cd
Here we have introduced a second External_interfaces.cfg  because there was no way to assure the
order of checkout of repositories in Externals.cfg, note also that issue #148 caused us to need to list
External_interfaces.cfg as an external of cime when I really wanted to name it as an external of ufs-s2s-model

If order were imposed on the Externals.cfg file all of these repos could be listed in a single externals file.

**Comment by jedwards4b** [comment]
If nesting repositories is not supported (checking out a repo within the directory structure of an existing repo) than that should explicitly cause an error condition.

**Comment by gold2718** [comment]
I guess I see this as a bonus in that the new CIME 'external' needs to be synced with the CIME version it is going to be inserted in. It can also be viewed as a limitation of CIME if CIME really requires drivers to be in a particular place in the CIME source tree (as opposed to something found in config_files.xml such as `COMP_ROOT_DIR_CPL`).

However, if you still want to implement this functionality, I will reiterate that I think this is a dramatic change for manage_externals semantics which can easily cause unexpected, negative side effects that would make manage_externals unusable by some current users. Will you take responsibility to test existing uses to make sure they still function properly in all conditions (e.g., with different levels of subrepo modification)?

**Comment by jedwards4b** [comment]
I don't understand your response, this has nothing to do with cime other than I've listed the interfaces as an External of cime because it wouldn't let me list them as an external of ufs-s2s-model.

The directory location is a noaa requirement not a cime one.  It doesn't need to make sense it just needs to work properly.   Can you give an example of how imposing an order on the Externals.cfg would corrupt the functionality of any repo?
If so then I would propose that an optional attribute be added to Externals.cfg:
ordered = True
if not provided ordered would default to False.

**Comment by gold2718** [comment]
If the default is to make no change, that should be okay _if this default is propagated to all parts of the code_.
Similarly, status checks could be affected if the non-default (ordered) semantics is not properly propagated to all parts of the code.

However, opinions aside, only thorough testing will determine whether this change does not break existing usage.

"... in theory there is no difference between theory and practice, while in practice there is" -- Benjamin Brewster, 1882

**Comment by jedwards4b** [comment]
I have no idea how to force order in this spaghetti code - do you?

**Comment by gold2718** [comment]
Nope and therein lies my concern. At the top level, there is `ExternalsDescription` based on `dict`. This could be changed to `OrderedDict`.  However, elsewhere, especially, in sourcetree.py, there are dictionaries all over the place and I have no idea how those need to be converted to make this work.
For instance, if there is a mix of required and optional externals, they can end up in different dictionaries and imposing an order becomes a nightmare.
Welcome to the rabbit hole.

**Comment by gold2718** [comment]
Here's an idea; Instead of imposing an ordering on the entire Externals file, how about adding a keyword such as `prerequisite` or `dependency`? That way, this could be checked before any repository action and ensure proper action ordering.

**Comment by uturuncoglu** [comment]
@gold2718 when I had a problem in UFS HAFS application and I tried to implement ordering but to read the External.cfg file an external Python module is used and that module does not support ordering. If we plan to add ordering, it must be after that step. BTW, I like your idea to add dependency. Maybe there is no need to order everything as you said.

**Comment by gold2718** [comment]
@uturuncoglu, the PR I made back to the UFS HAFS application should fix your issues with `checkout_externals`. When I tried it, it worked quite easily.
